### PR TITLE
Implement presto_typeof function for discovering variable type.

### PR DESCRIPTION
**Pull request by fiedukow**
The function may be usefull for debugging for presto developers
as well as for presto user while creating complex SQL queries.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.type.SqlType;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.nio.ByteBuffer;
+
+@Description("Return type of scalar parameter as VARCHAR")
+@ScalarFunction("presto_typeof")
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:26]: why `presto_typeof` and not just `typeof` or `type`?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import com.facebook.presto.operator.Description;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.type.SqlType;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.nio.ByteBuffer;
+
+@Description("Return type of scalar parameter as VARCHAR")
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:25]: `"Return textual representation of expression passed as a parameter"`?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+@ScalarFunction("presto_typeof")
+public final class PrestoTypeOfFunction
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:27]: Same here - drop `Presto` prefix if possible.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+{
+    private PrestoTypeOfFunction() {}
+
+    @TypeParameter("T")
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice typeof(
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:33]: does this work? I am not sure how functions are matched based on container types. If  supertype is allowed?

Please add some tests.
For different types including maps, arrays.

**Comment by losipiuk** [comment]
looks good but please add some tests.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java (hunk 1 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:26]: It is similar to Postgree `pg_typeof`. I'm not sure why did thay add such prefix, but i have some doubts that it is for future compatibility, if at some point similar function will be introduced in ANSI-SQL standard.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java (hunk 4 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:33]: As far as i know it should be matched without any issues. But you are right, it is always better to add a tests.

**Comment by arhimondr** [comment]
Nice. Now we don't need to create temporary tables in order to determine the expression output type ;-)

**Comment by ilfrin** [comment]
closing this, reopen if needed. In general we should not make PRs to Teradata/presto:master but prestodb/presto:master. If making PR to Teradata fork it should be some development branch of ours (eg. sprint-*). Teradata/presto:master is "read-only".

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            @TypeParameter("T") Type type,
+            @SqlType("T") Object value)
+    {
+        return Slices.wrappedBuffer(ByteBuffer.wrap(type.toString().getBytes()));
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/PrestoTypeOfFunction.java:37]: `getDisplayName()` instead of `toString()` maybe?

**Pull request closed without merging**